### PR TITLE
Add core component upgrade design doc

### PR DESCRIPTION
**Pull request by g-gaston**
*Description of changes:*
Design doc to upgrade cluster components

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/296#" title="Author self-approved">g-gaston</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-e2e-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-e2e-presubmit/1441123277266227200) | `/test eks-anywhere-e2e-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=296). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-e2e-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-e2e-presubmit/1441123277266227200) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-mocks-presubmits | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-mocks-presubmits/1441123277299781632) | `/test eks-anywhere-mocks-presubmits`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=296). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-e2e-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-e2e-presubmit/1441123277266227200) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-mocks-presubmits | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-mocks-presubmits/1441123277299781632) | `/test eks-anywhere-mocks-presubmits`
eks-anywhere-attribution-files-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-attribution-files-presubmit/1441123277236867072) | `/test eks-anywhere-attribution-files-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=296). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-e2e-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-e2e-presubmit/1441123277266227200) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-mocks-presubmits | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-mocks-presubmits/1441123277299781632) | `/test eks-anywhere-mocks-presubmits`
eks-anywhere-attribution-files-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-attribution-files-presubmit/1441123277236867072) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-presubmit | dac5dbe955008a3a96fca453c68a7bcf04ed0380 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/296/eks-anywhere-presubmit/1441123277329141760) | `/test eks-anywhere-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=296). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:131]: Let's say we roll out a change and want to revert back to an older release. Will customers be able to run upgrade cluster to get that out or will it always be a roll-forward in our case? For EKS, there are automatic updates, but I believe we also have the case where changes can get "rolled-back" if there is an issue automatically, so curious about the behavior here.

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:137]: We should mention that this is similar to how EKS handles auto updating, where auto updating here is handled manually through the upgrade command

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:36]: This will be in scope whenever we have add ons that are part of the create workflow right? Unless we have documentation mentioning otherwise.

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:26]: Is users specifying an override bundle in scope?

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:62]: I would call out that another benefit is that customers technically never have to worry about the individual version and dependencies, which also a benefit of this project.

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:8]:
```suggestion
This stops users from enjoying the latest features and bug fixes from eks-a and any of its core components as well as from being able to get security patches.
```

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:16]:
```suggestion
As an eks-a user:
```

We should just reference users everywhere else as well

**Comment by rothgar** [comment]
Will `cluster upgrade` have an option for a `--dry-run` to fetch the latest bundle and show a user what will change?

There also should be some information about skipping versions (eg 1.21 -> 1.23) and let users know (and build it into the CLI) that only 1 point release is possible for each `upgrade` command.

Do we have any ideas for having subscriptions for releases? GKE has [release channels](https://cloud.google.com/kubernetes-engine/docs/concepts/release-channels) which allow automatic upgrades without user intervention. It would be a good user experience to provide a similar mechanism in EKS-A clusters defined in the cluster spec.

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:76]: We also need to cross-reference with what is already installed, meaning having it in our Cluster status for example on what bundle is currently being consumed for the cluster. Sounds like we need to pick this issue up for that then? https://github.com/aws/eks-anywhere/issues/165

File: designs/component-upgrades.md
Comment (vivek-koppuru) [inline designs/component-upgrades.md:139]: These should also include error cases that we throw mentioning that a user needs to upgrade CLI/Kubernetes version right?

**Review by vivek-koppuru — COMMENTED** [review]
This will definitely help make upgrading components seamless 😄

**Comment by vivek-koppuru** [comment]
> Will `cluster upgrade` have an option for a `--dry-run` to fetch the latest bundle and show a user what will change?
> 
> There also should be some information about skipping versions (eg 1.21 -> 1.23) and let users know (and build it into the CLI) that only 1 point release is possible for each `upgrade` command.
> 
> Do we have any ideas for having subscriptions for releases? GKE has [release channels](https://cloud.google.com/kubernetes-engine/docs/concepts/release-channels) which allow automatic upgrades without user intervention. It would be a good user experience to provide a similar mechanism in EKS-A clusters defined in the cluster spec.

The release channels concept sounds interesting if we can have the Kubernetes cluster wait on that, but that is also a cloud benefit and could be implemented as a followup once we have this down.

**Comment by g-gaston** [comment]
> Will `cluster upgrade` have an option for a `--dry-run` to fetch the latest bundle and show a user what will change?
> 
> There also should be some information about skipping versions (eg 1.21 -> 1.23) and let users know (and build it into the CLI) that only 1 point release is possible for each `upgrade` command.
> 
> Do we have any ideas for having subscriptions for releases? GKE has [release channels](https://cloud.google.com/kubernetes-engine/docs/concepts/release-channels) which allow automatic upgrades without user intervention. It would be a good user experience to provide a similar mechanism in EKS-A clusters defined in the cluster spec.

Showing prospected changes, subscriptions/notifications and automatic/scheduled upgrades are part of the future scope, but not in scope for this doc/project. Let me know if that section of the doc needs a bit more clarification.

That said, something like `--dry-run` or a `upgrade cluster plan` command shouldn't be very difficult to implement. I think it could easily be a fast follower to this project, good idea.

Regarding skipping kubernetes versions, I believe that's already handled by the current `upgrade cluster` command implementation. There are validations in place and it should print the proper error. I think it's also mentioned in the docs. Or are you thinking about something else?

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:137]: you mean add that to the docs?

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:88]:
```suggestion
We have to be careful and make sure that any changes to the CRDs are backwards compatible, since at this point there are already EKS Anywhere objects in the cluster. Examples of changes that are not backwards compatible include: adding a required field, changing the type of field, making an existing field required, and so on. It’s important to note that applying the manifest with the new CRD won’t fail, even if the changes are not backwards compatible. But the objects already in the cluster will become invalid. So any kind of update to those (such as flux reconciliation) will be be rejected by the Kubernetes API server.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:92]:
```suggestion
If there is a new EKS Anywhere API version, we will need to write conversion webhooks.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:96]:
```suggestion
The process to upgrade the Flux components is very similar to the installation process. It uses the `flux` cli:
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:108]:
```suggestion
The process to upgrade Cilium is a bit more involved. It uses `helm` to generate intermediate manifests and `kubectl` to apply them.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:110]:
```suggestion
1. Generate pre-flight checks with `helm` and apply them with `kubectl`. This “pre-pulls" the new images on the nodes to avoid `ErrImagePull` errors later during the upgrade
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:111]:
```suggestion
2. Once it’s finished, delete the pre-flight components with `kubectl`
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:112]:
```suggestion
3. Generate the new Cilium manifest using the right “upgrade compatibility” options with `helm` (the generated manifest is different depending on the minor version jump)  and apply it with `kubectl`.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:120]:
```suggestion
As mentioned in the [CAPI](#capi) section, we don’t need any extra process to perform CAPI component upgrades that involve an API version change. However, there will some extra effort needed to support `v1alpha4` in the cli. Most of this will probably apply to future API version changes.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:131]:
```suggestion
There are not backwards breaking changes. However, the expected behavior of the `upgrade cluster` command will be changing, so it should be clearly communicated in the changelog.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:137]:
```suggestion
In the `upgrade cluster` command documentation we should document that we will always upgrade internal EKS Anywhere components to the latest available version, even when the Kubernetes version remains unchanged.
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:144]:
```suggestion
    * For `main`, this makes sure that the next (future) minor release is compatible with clusters created by the current latest release
```

File: designs/component-upgrades.md
Comment (chrisnegus) [inline designs/component-upgrades.md:145]:
```suggestion
    * When running this in a release branch, this makes sure that the next (future) patch release is compatible with clusters created by the previous minor release
```

**Review by chrisnegus — COMMENTED** [review]
Good document. I learned a lot. Most of my suggestions are just style and terminology issues.

File: designs/component-upgrades.md
Comment (abhay-krishna) [inline designs/component-upgrades.md:32]: Should we also include cert-manager in this list?

File: designs/component-upgrades.md
Comment (abhay-krishna) [inline designs/component-upgrades.md:68]:
```suggestion
Components should be **upgraded** **before the new Kubernetes version is rolled out** (if needed). It is more common for an old component to be incompatible with a new Kubernetes version than vice-versa. There might be case by case exceptions, right now there is not, at least with the current set of core components.
```

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:131]: Great question
The answer depends on the what's the version change (minor, patch, api version bump, just a new build...).
For example, if we are just reverting back a new build of the components version, yes, we only need a new bundles manifest that points to the previous build. When customers run the upgrade command, the "old" version would be installed.
But, even if this is a downgrade, it's not a rollback per se, it's more a roll-forward like you mentioned. A proper rollback would take the cluster back to the previous installed version, which might not be the version from the previous bundle manifest (image a customer skipping one version upgrade).

For more "drastic" upgrades, like something that involves a CAPI API bump, we would need to test it case by case. Also that would involve a new cli release, since we would need to revert the providers' templates as well.

To be clear, this design assumes that the version changes are always forward. If we want the ability to downgrade versions, we would need to come up with a process for each component (the process for upgrade might work here, but it would need to be tested). If we want the ability to do proper roll-backs, then that also would need a bit more research, since most probably each component would require a slightly different process.

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:36]: I'm not sure we have such documentation
However, I was thinking that add-ons might have their own lifecycle system, whether they are upgraded through the `upgrade cluster` command or not. So I considered their upgrade flow a different feature, so not in scope for this one.

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:26]: I wouldn't say so
However, with the implementation I have in mind that should work out of the box. But I don't think we should make it part of the scope

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:76]: That's a good point
However, it's not completely necessary. The current controller is capable of retrieving the installed `VersionsBundle`, so the cli should be able to do the same (following the same naming assumptions)

That said, I think using the status is a way better solution. So if we have bandwidth, it would be ideal to make it part of this project. Maybe we can implement a centralized logic to get the "current `VersionBundle`" and use it from the cli and controller. And whenever we pick that ticket, just update that place, whether is a part of this project or later.

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:139]: I'm not sure exactly what you mean
Could you give one concrete example of the scenario?

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:32]: Good idea
I will also mention it in the CAPI upgrade section, since `clusterctl` will update it as well

File: designs/component-upgrades.md
Comment (vivek-koppuru) [reply designs/component-upgrades.md]: I just mean as a distinction in the design here, because that's another reason we want to keep it simple like this right?

File: designs/component-upgrades.md
Comment (vivek-koppuru) [reply designs/component-upgrades.md:37]: So is Flux the only exception here?

File: designs/component-upgrades.md
Comment (vivek-koppuru) [reply designs/component-upgrades.md:141]: If a user specifies a kubernetes version that is too far out, we would throw an error saying it's not supported and that maybe they would have to upgrade their CLI version.

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md]: Added to project task list

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:141]: Note: we already have unit tests for this

File: designs/component-upgrades.md
Comment (g-gaston) [reply designs/component-upgrades.md:37]: Yeah Flux is the exception, it's not a critical component but it's a core component of the eks-a experience

**Review by vivek-koppuru — APPROVED** [review]
/lgtm

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/296#" title="Author self-approved">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/296#pullrequestreview-764751607" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
